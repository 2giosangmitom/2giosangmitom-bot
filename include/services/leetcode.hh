#pragma once

#include <nlohmann/json.hpp>
#include <optional>
#include <set>
#include <string>
#include <vector>

namespace leetcode {

// Represents a single LeetCode problem record.
struct LeetCodeProblem {
  int id;                 // Unique identifier for the problem.
  std::string title;      // Full title of the problem.
  std::string difficulty; // Difficulty level ("Easy", "Medium", "Hard").
  std::vector<std::string>
      topics;      // List of topics/tags related to the problem.
  std::string url; // Web URL to the problem statement.
  bool is_paid;    // True if the problem is paid-only.
  double ac_rate;  // Acceptance rate as a percentage.
};

// Stores metadata and a collection of LeetCode problems.
struct Data {
  std::vector<LeetCodeProblem> problems; // List of all LeetCode problems.
  std::set<std::string> topics; // Set of all unique topics across problems.
  std::string lastUpdated;      // Timestamp of the last data update.
  int totalProblems;            // Total number of problems included.
};

NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(LeetCodeProblem, id, title, difficulty,
                                   topics, url, is_paid, ac_rate);
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Data, problems, topics, lastUpdated,
                                   totalProblems);

// Downloads the latest data from the LeetCode GraphQL API and caches it
// to 'data.json'.
bool download_data();

// Validate the json data
bool validate_json(const nlohmann::json &json_data);

// Read data.json file
Data load_data_json();

// Filters LeetCode problems by difficulty and/or topics.
std::vector<LeetCodeProblem>
filter_questions(const std::optional<std::vector<std::string>> &difficulty,
                 const std::optional<std::vector<std::string>> &topics);

// Selects a random subset of LeetCode problems from a given collection.
std::vector<LeetCodeProblem>
get_questions(const std::vector<LeetCodeProblem> &problems,
              std::optional<int> quantity);

} // namespace leetcode