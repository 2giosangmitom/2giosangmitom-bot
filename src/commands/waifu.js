import { SlashCommandBuilder, EmbedBuilder } from 'discord.js';
import { getImage, categories } from '../services/waifu.js';
import { randomFrom } from '../lib/utils.js';

/** @type {string[]} */
const titles = [
  "Here's Your Daily Dose of Motivation ✨",
  'A Waifu Appears! 💖',
  'Stay Strong, Senpai! 💪',
  "You Got This! Here's Some Motivation 🔥",
  'Summoning Your Waifu... 💫',
  'Your Waifu Believes in You! 🌸',
  'Power Up Time! 🚀',
  "Don't Give Up, Senpai! 💥",
  'Waifu Buff Activated! ⚡',
  "Keep Going, You're Doing Great! 🌟",
  'One Step Closer to Victory! 🏆',
  'Level Up Your Spirit! 🆙',
  'Another Day, Another Quest! 🗺️',
  "You're Stronger Than You Think! 🐉",
  "Waifu's Blessing Incoming! 🍀"
];

const data = new SlashCommandBuilder()
  .setName('waifu')
  .setDescription('Get a random cute anime girl image to boost your motivation 💖')
  .addStringOption((option) =>
    option
      .setName('category')
      .setDescription('The category of image')
      .addChoices(categories.map((v) => ({ name: v, value: v })))
  );

/**
 * Executes the waifu command and returns a motivational anime image
 * @param {import('discord.js').ChatInputCommandInteraction} interaction - The command interaction
 * @returns {Promise<void>}
 */
async function execute(interaction) {
  await interaction.deferReply();

  try {
    const categoryParam = interaction.options.getString('category')?.toLowerCase() || undefined;
    const { url, category } = await getImage(categoryParam);

    const embed = new EmbedBuilder()
      .setColor(15431372)
      .setTitle(randomFrom(titles) || 'Your Waifu is Here! 💖')
      .setDescription(`*Category: ${category}*`)
      .setImage(url)
      .setFooter({ text: 'Powered by waifu.pics' })
      .setTimestamp();

    await interaction.editReply({ embeds: [embed] });
  } catch (error) {
    throw new Error(`Failed to fetch waifu image: ${error instanceof Error ? error.message : 'Unknown error'}`);
  }
}

export { data, execute };
