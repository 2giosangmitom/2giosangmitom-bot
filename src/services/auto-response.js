import { randomFrom } from '../lib/utils.js';

/**
 * @typedef Trigger
 * @property {RegExp} pattern - The regex pattern to match normalized message content
 * @property {string[]} responses - A list of possible replies
 */

/** @type {Trigger[]} */
const triggers = [
  {
    pattern: /\b(vl|vcl|dcm|dm)\b/i,
    responses: ['Chui tuc con cark 🚫']
  },
  {
    pattern: /hello|hi|chao|yo|xin chao/i,
    responses: ['Chao ban 👋', 'Hi hi 😄', 'Toi nghe day!', 'Hello hello!']
  },
  {
    pattern: /cho tao xem meme/i,
    responses: ['Meme con cark', 'Xem meme lam j', 'Xem cuc cut']
  }
];

/**
 * Normalize Vietnamese: remove diacritics and convert to lowercase
 * Example: "Đây là tiếng Việt!" -> "day la tieng viet!"
 * @param {string} str - The string that needs to be normalized
 * @returns {string}
 */
function normalize(str) {
  return str
    .normalize('NFD') // Decompose letters and accents
    .replace(/[\u0300-\u036f]/g, '') // Remove diacritic marks
    .replace(/đ/g, 'd') // Convert đ to d
    .replace(/Đ/g, 'D')
    .toLowerCase();
}

/**
 * Automatically generates a response based on message content
 * @param {string} rawContent - Raw message content from Discord
 * @returns {string | null} A bot response, or null if no trigger matches
 */
function response(rawContent) {
  const content = normalize(rawContent);

  for (const trigger of triggers) {
    if (trigger.pattern.test(content)) {
      return randomFrom(trigger.responses);
    }
  }

  return null;
}

export { response, normalize };
