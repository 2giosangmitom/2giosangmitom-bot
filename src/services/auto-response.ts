/**
 * @author Vo Quang Chien <voquangchien.dev@proton.me>
 * @license MIT
 * @copyright Â© 2025 Vo Quang Chien
 */

import { randomFrom } from '~/lib/utils';

interface Trigger {
  description: string; // Description for the trigger
  pattern: RegExp; // RegExp to check the message content
  responses: string[]; // Response messages
}

/** @description Regexes for checking whether the bot should reply to the user */
const triggers: Trigger[] = [
  {
    description: 'Greeting',
    pattern: /\b(hello|hi|chao|yo)\b|xin chao/i,
    responses: ['Chao ban ðŸ‘‹', 'Hi hi ðŸ˜„', 'Toi nghe day!', 'Hello hello!']
  }
];

/**
 * @description Get normalized string, remove diacritics and convert to lowercase.
 * @param str The string to normalize
 */
function normalize(str: string) {
  return str
    .normalize('NFD') // Decompose letters and accents
    .replace(/[\u0300-\u036f]/g, '') // Remove diacritic marks
    .replace(/Ä‘/gi, 'd') // Convert Ä‘ to d
    .toLowerCase();
}

/**
 * @description Automatically generates a response based on message content.
 * @param content Message content from Discord
 */
function replyMessage(content: string): string | null {
  content = normalize(content);

  for (const trigger of triggers) {
    if (trigger.pattern.test(content)) {
      return randomFrom(trigger.responses);
    }
  }

  return null;
}

export { triggers, replyMessage, normalize };
