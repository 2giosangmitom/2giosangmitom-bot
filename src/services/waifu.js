import { randomFrom } from '../lib/utils.js';
import * as https from 'node:https';

const categories = ['waifu', 'hug', 'kiss', 'happy', 'handhold', 'bite'];
const url = 'https://api.waifu.pics/sfw';

/**
 * Calls the waifu.pics API and returns an image URL with its category.
 * @param {string} [category] - Optional category to fetch; will pick random if undefined
 * @returns {Promise<{ url: string, category: string }>}
 */
function getImage(category) {
  return new Promise((resolve, reject) => {
    if (!category) {
      category = randomFrom(categories);
    }
    if (!categories.includes(category)) {
      return reject('The requested category is not available');
    }

    https
      .get(`${url}/${category}`, (res) => {
        const { statusCode, headers } = res;

        if (statusCode !== 200) {
          res.resume(); // Discard response to free memory
          return reject(`Request Failed. Status Code: ${statusCode}`);
        }

        if (!/^application\/json/.test(headers['content-type'])) {
          res.resume();
          return reject(`Invalid content-type. Expected application/json but received ${headers['content-type']}`);
        }

        let rawData = '';
        res.setEncoding('utf-8');
        res.on('data', (chunk) => (rawData += chunk));

        res.on('end', () => {
          try {
            const json = JSON.parse(rawData);
            resolve({ url: json?.url, category });
          } catch (e) {
            reject(e);
          }
        });
      })
      .on('error', (e) => {
        reject(e);
      });
  });
}

export { getImage, categories };
