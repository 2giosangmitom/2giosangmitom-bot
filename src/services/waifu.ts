import { randomFrom } from '../lib/utils.js';
import { inlineCode } from 'discord.js';
import z from 'zod';

namespace WaifuService {
  // Allowed categories from waifu.pics
  export const categories = ['waifu', 'hug', 'kiss', 'happy', 'handhold', 'bite', 'slap'];

  // Titles for the embed message
  export const titles = [
    "Here's Your Daily Dose of Motivation ✨",
    'A Waifu Appears! 💖',
    'Stay Strong, Senpai! 💪',
    "You Got This! Here's Some Motivation 🔥",
    'Summoning Your Waifu... 💫',
    'Your Waifu Believes in You! 🌸',
    'Power Up Time! 🚀',
    "Don't Give Up, Senpai! 💥",
    'Waifu Buff Activated! ⚡',
    "Keep Going, You're Doing Great! 🌟",
    'One Step Closer to Victory! 🏆',
    'Level Up Your Spirit! 🆙',
    'Another Day, Another Quest! 🗺️',
    "You're Stronger Than You Think! 🐉",
    "Waifu's Blessing Incoming! 🍀",
    "Even Mondays Can't Stop You! ☕",
    'Waifu Says: Believe in Yourself! 🦸',
    'You Leveled Up in Awesomeness! 🎮',
    'No Boss Fight Can Beat You! 🗡️',
    'Waifu Cheering From the Sidelines! 📣',
    'Critical Hit of Positivity! 💥',
    'Your Effort is Super Effective! 🧩',
    'Waifu Sent You Extra Luck Today! 🍀',
    'Keep Grinding, Hero! 💎',
    "You're the Main Character Today! 🎬",
    "Waifu's Power-Up: Unlimited Motivation! 🔋",
    'Achievement Unlocked: Keep Going! 🏅',
    "Waifu's Smile Restores 100 HP! 💚",
    'Plot Armor Activated! 🛡️',
    "You're Destined for Greatness! 🌠"
  ];

  export async function getImage(
    category?: string
  ): Promise<{ url: string; category: string; title: string }> {
    category = category ?? randomFrom(categories)!;

    // Reject if the provided category is not valid
    if (!categories.includes(category)) {
      throw new Error(`The ${inlineCode(category)} category is not valid`);
    }

    // Fetch the image
    const res = await fetch(`https://api.waifu.pics/sfw/${category}`);

    if (!res.ok) {
      throw new Error(
        `The waifu.pics API is not available at the moment. Status code: ${res.status}`
      );
    }

    // Validate the response JSON
    const schema = z.object({
      url: z.url()
    });

    const result = schema.safeParse(await res.json());
    if (!result.success) {
      throw new Error(
        `The responsed JSON from waifu.pics is not valid.\n${z.prettifyError(result.error)}`
      );
    }

    return {
      url: result.data.url,
      category,
      title: randomFrom(titles)!
    };
  }
}

export default WaifuService;
