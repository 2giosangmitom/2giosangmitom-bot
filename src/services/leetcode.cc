#include "services/leetcode.hh"
#include "utils/random.hh"
#include "utils/string.hh"
#include <algorithm>
#include <cerrno>
#include <cstring>
#include <curl/curl.h>
#include <fstream>
#include <nlohmann/json.hpp>
#include <spdlog/spdlog.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

// Constants
static constexpr const char *leetcode_api_url = "https://leetcode.com/graphql";
static constexpr const char *output_file = "data.json";
const std::vector<std::string> leetcode::difficulties{"Easy", "Medium", "Hard"};

using leetcode::LeetCodeProblem;
using nlohmann::json;

// Helper function for libcurl to write received data into a std::ostream
size_t write_cb(void *contents, size_t size, size_t nmemb, void *userp) {
  std::ostream *stream = static_cast<std::ostream *>(userp);
  size_t realsize = size * nmemb;
  stream->write(static_cast<char *>(contents), realsize);
  return realsize;
}

// Helper function for building payload data
inline json build_graphql_payload() {
  const std::string query = R"(
query problemsetQuestionListV2($filters: QuestionFilterInput, $limit: Int, $skip: Int) {
  problemsetQuestionListV2(
    filters: $filters
    limit: $limit
    skip: $skip
  ) {
    questions {
      id
      titleSlug
      title
      questionFrontendId
      paidOnly
      difficulty
      topicTags {
        name
        slug
      }
      acRate
    }
    totalLength
  }
}
  )";

  return {
      {"query", query},
      {"variables",
       {
           {"skip", 0},
           {"limit", 10000},
           {"filters", {{"filterCombineType", "ALL"}}},
       }},
  };
}

// Helper function for caching API response to "data.json"
bool write_json_to_file(const json &json) {
  std::ofstream out(output_file, std::ios::out | std::ios::trunc);
  if (!out.is_open()) {
    spdlog::critical("Failed to open '{}' for writing: {}", output_file,
                     strerror(errno));
    return false;
  }

  out << json.dump(2) << std::endl;
  if (!out) {
    spdlog::critical("Failed to write to '{}': {}", output_file,
                     strerror(errno));
    return false;
  }

  return true;
}

bool leetcode::validate_json(const json &json_data) {
  // Check if "metadata" exists and is an object
  const auto meta_it = json_data.find("metadata");
  if (meta_it == json_data.end() || !meta_it->is_object()) {
    return false;
  }

  // Check required metadata fields
  if (!meta_it->contains("lastUpdated") || !meta_it->contains("totalProblems"))
    return false;
  if (!(*meta_it)["lastUpdated"].is_string() ||
      !(*meta_it)["totalProblems"].is_number())
    return false;

  // Check "problems" is present and is an array
  const auto problems_it = json_data.find("problems");
  if (problems_it == json_data.end() || !problems_it->is_array())
    return false;

  // Check each problem has required structure
  for (const auto &problem : *problems_it) {
    if (!problem.contains("id") || !problem["id"].is_number())
      return false;
    if (!problem.contains("title") || !problem["title"].is_string())
      return false;
    if (!problem.contains("difficulty") || !problem["difficulty"].is_string())
      return false;
    if (!problem.contains("url") || !problem["url"].is_string())
      return false;
    if (!problem.contains("is_paid") || !problem["is_paid"].is_boolean())
      return false;
    if (!problem.contains("ac_rate") || !problem["ac_rate"].is_number())
      return false;
    if (!problem.contains("topics") || !problem["topics"].is_array())
      return false;
  }

  // Check "topics" array exists and is an array
  if (json_data.contains("topics") && !json_data["topics"].is_array())
    return false;

  return true;
}

bool leetcode::download_data() {
  spdlog::info("Preparing LeetCode GraphQL request...");

  auto payload = build_graphql_payload();
  const std::string graphql_payload = payload.dump();

  CURL *handle = curl_easy_init();
  if (!handle) {
    spdlog::critical("Failed to initialize curl handle.");
    return false;
  }

  curl_slist *headers = nullptr;
  headers = curl_slist_append(headers, "Content-Type: application/json");

  std::stringstream response;
  curl_easy_setopt(handle, CURLOPT_URL, leetcode_api_url);
  curl_easy_setopt(handle, CURLOPT_HTTPHEADER, headers);
  curl_easy_setopt(handle, CURLOPT_POSTFIELDS, graphql_payload.c_str());
  curl_easy_setopt(handle, CURLOPT_POSTFIELDSIZE, graphql_payload.size());
  curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, write_cb);
  curl_easy_setopt(handle, CURLOPT_WRITEDATA, &response);

  spdlog::info("Sending request to LeetCode API...");
  CURLcode res = curl_easy_perform(handle);
  curl_slist_free_all(headers);
  curl_easy_cleanup(handle);

  if (res != CURLE_OK) {
    spdlog::critical("curl_easy_perform failed: {}", curl_easy_strerror(res));
    return false;
  }

  json json_response;
  try {
    json_response = json::parse(response.str());
  } catch (const std::exception &e) {
    spdlog::critical("Failed to parse LeetCode response JSON: {}", e.what());
    return false;
  }

  auto &questions =
      json_response["data"]["problemsetQuestionListV2"]["questions"];
  if (!questions.is_array()) {
    spdlog::critical("Unexpected structure in response.");
    return false;
  }

  spdlog::info("Parsed {} problems.", questions.size());

  // Convert to internal format
  Data data;
  for (const auto &q : questions) {
    LeetCodeProblem problem;
    problem.id = q.value("id", 0);
    problem.title = q.value("title", "Unknown");
    problem.difficulty = q.value("difficulty", "Unknown");
    problem.is_paid = q.value("paidOnly", false);
    problem.ac_rate = q.value("acRate", 0.0);
    problem.url =
        "https://leetcode.com/problems/" + q.value("titleSlug", "") + "/";

    for (const auto &tag : q["topicTags"]) {
      if (tag.contains("name")) {
        std::string topic = tag["name"];
        problem.topics.push_back(topic);
        data.topics.insert(topic);
      }
    }

    data.problems.push_back(std::move(problem));
  }

  data.metadata.totalProblems = data.problems.size();
  data.metadata.lastUpdated = string_utils::get_timestamp("%d %B");

  json final_json = {
      {"metadata",
       {
           {"lastUpdated", data.metadata.lastUpdated},
           {"totalProblems", data.metadata.totalProblems},
       }},
      {"problems", data.problems},
      {"topics", data.topics},
  };

  if (!write_json_to_file(final_json)) {
    return false;
  }

  spdlog::info("LeetCode data written to '{}'.", output_file);
  return true;
}

leetcode::Data leetcode::load_data_json() {
  std::ifstream in(output_file);
  if (!in.is_open()) {
    throw std::runtime_error(
        fmt::format("Failed to open data.json: {}", strerror(errno)));
  }

  json data_json;

  try {
    in >> data_json;
  } catch (const json::parse_error &e) {
    throw std::runtime_error(
        fmt::format("Failed to parse data.json: {}", e.what()));
  }

  if (!validate_json(data_json)) {
    throw std::runtime_error("Invalid format for \"data.json\"");
  }

  return data_json.get<leetcode::Data>();
}

std::vector<LeetCodeProblem> leetcode::filter_questions(
    const std::vector<LeetCodeProblem> &problems,
    std::optional<std::vector<std::string>> &difficulties,
    std::optional<std::vector<std::string>> &topics) {
  std::vector<LeetCodeProblem> filtered;

  for (auto &q : problems) {
    if (difficulties.has_value() &&
        !std::any_of(difficulties->begin(), difficulties->end(),
                     [&q](std::string &a) { return a == q.difficulty; })) {
      continue;
    }

    if (topics.has_value() &&
        !std::any_of(topics->begin(), topics->end(), [&q](std::string &a) {
          return std::any_of(
              q.topics.begin(), q.topics.end(),
              [&a](std::string &q_topic) { return q_topic == a; });
        })) {
      continue;
    }

    filtered.push_back(q);
  }

  return filtered;
}

std::vector<LeetCodeProblem>
leetcode::get_questions(const std::vector<LeetCodeProblem> &filtered_problems,
                        std::optional<int> quantity) {
  if (!quantity.has_value()) {
    quantity = 1; // Default value
  }
  if (quantity.value() > 10) {
    quantity = 10; // Limit to 10 questions
  }

  std::vector<LeetCodeProblem> result;
  int total = filtered_problems.size();

  for (int i = 1; i <= quantity.value(); i++) {
    int rand_idx = random_utils::random_int_range(0, total - 1);
    result.push_back(filtered_problems[rand_idx]);
  }

  return result;
}