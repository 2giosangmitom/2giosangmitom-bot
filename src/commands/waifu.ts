/**
 * @file Waifu command
 * @author Vo Quang Chien <voquangchien.dev@proton.me>
 */

import { ChatInputCommandInteraction, EmbedBuilder, italic, SlashCommandBuilder } from 'discord.js';
import { getImage, categories } from '~/services/waifu';
import type { SlashCommand } from '~/types';

const command: SlashCommand = {
  data: new SlashCommandBuilder()
    .setName('waifu')
    .setDescription('Get a random cute anime girl image to boost your motivation ðŸ’–')
    .addStringOption((option) =>
      option
        .setName('category')
        .setDescription('The category of the image')
        .addChoices(categories.map((category) => ({ name: category, value: category })))
    ),
  async execute(interaction: ChatInputCommandInteraction) {
    await interaction.deferReply();

    const categoryParam = interaction.options.getString('category') ?? undefined;

    const { url, category, title } = await getImage(categoryParam);

    const waifuEmbed = new EmbedBuilder()
      .setColor('LuminousVividPink')
      .setTitle(title)
      .setDescription(italic(`Category: ${category}`))
      .setImage(url)
      .setFooter({ text: 'Powered by waifu.pics' })
      .setTimestamp();

    await interaction.followUp({ embeds: [waifuEmbed] });
  }
};

export default command;
