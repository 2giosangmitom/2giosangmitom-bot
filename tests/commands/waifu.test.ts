/**
 * @file Unit tests for waifu command
 * @author Vo Quang Chien <voquangchien.dev@proton.me>
 */

import { afterEach, beforeEach, describe, it, vi, expect } from 'vitest';
import type { MockChatInteraction } from '~/types';
import * as waifu from '~/services/waifu';
import waifuCommand from '~/commands/waifu';

describe('waifu command', () => {
  let mockInteraction: MockChatInteraction;

  beforeEach(() => {
    mockInteraction = {
      deferReply: vi.fn().mockResolvedValue(true),
      options: {
        getString: vi.fn().mockReturnValue('mock')
      },
      followUp: vi.fn()
    } as MockChatInteraction;
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it.each([
    { url: 'https://waifu.mock', category: 'hehe', title: 'oidoioi' },
    { url: 'https://mock.waifu', category: 'waifu', title: 'i love you, Truyen' },
    { url: 'https://cdn.waifu.mock/img1234567890.png', category: '', title: 'Untitled Waifu' },
    { url: 'https://waifu.mock/nekos', category: 'nekos', title: 'Classic Neko Style' }
  ])(
    'should reply to user correctly with url = $url, category = $category, title = $title',
    async ({ url, category, title }) => {
      const spyGetImage = vi.spyOn(waifu, 'getImage');
      spyGetImage.mockResolvedValue({
        url,
        category,
        title
      });
      vi.setSystemTime(Date.UTC(2022, 0, 15, 17, 15, 20, 66));

      await waifuCommand.execute(mockInteraction);

      expect(mockInteraction.deferReply).toHaveBeenCalledOnce();
      expect(mockInteraction.options.getString).toHaveBeenCalledOnce();
      expect(spyGetImage).toHaveBeenCalledOnce();
      expect(mockInteraction.followUp.mock.calls[0]?.[0]).toMatchSnapshot();
    }
  );

  it('should work when category is not provided', async () => {
    mockInteraction.options.getString = vi.fn().mockReturnValue(null);
    const spyGetImage = vi.spyOn(waifu, 'getImage').mockResolvedValue({
      url: 'https://waifu.mock/default.jpg',
      category: 'default',
      title: 'default title'
    });

    await waifuCommand.execute(mockInteraction);

    expect(spyGetImage).toHaveBeenCalledWith(undefined);
    expect(mockInteraction.followUp).toHaveBeenCalledOnce();
  });
});
