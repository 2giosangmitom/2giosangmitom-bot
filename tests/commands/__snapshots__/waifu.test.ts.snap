// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`waifu command > should reply to user correctly with url = 'https://cdn.waifu.mock/img1234567890.â€¦', category = '', title = 'Untitled Waifu' 1`] = `
{
  "embeds": [
    {
      "color": 15277667,
      "description": "_Category: _",
      "footer": {
        "icon_url": undefined,
        "text": "Powered by waifu.pics",
      },
      "image": {
        "url": "https://cdn.waifu.mock/img1234567890.png",
      },
      "timestamp": "2022-01-15T17:15:20.066Z",
      "title": "Untitled Waifu",
    },
  ],
}
`;

exports[`waifu command > should reply to user correctly with url = 'https://mock.waifu', category = 'waifu', title = 'i love you, Truyen' 1`] = `
{
  "embeds": [
    {
      "color": 15277667,
      "description": "_Category: waifu_",
      "footer": {
        "icon_url": undefined,
        "text": "Powered by waifu.pics",
      },
      "image": {
        "url": "https://mock.waifu",
      },
      "timestamp": "2022-01-15T17:15:20.066Z",
      "title": "i love you, Truyen",
    },
  ],
}
`;

exports[`waifu command > should reply to user correctly with url = 'https://waifu.mock', category = 'hehe', title = 'oidoioi' 1`] = `
{
  "embeds": [
    {
      "color": 15277667,
      "description": "_Category: hehe_",
      "footer": {
        "icon_url": undefined,
        "text": "Powered by waifu.pics",
      },
      "image": {
        "url": "https://waifu.mock",
      },
      "timestamp": "2022-01-15T17:15:20.066Z",
      "title": "oidoioi",
    },
  ],
}
`;

exports[`waifu command > should reply to user correctly with url = 'https://waifu.mock/nekos', category = 'nekos', title = 'Classic Neko Style' 1`] = `
{
  "embeds": [
    {
      "color": 15277667,
      "description": "_Category: nekos_",
      "footer": {
        "icon_url": undefined,
        "text": "Powered by waifu.pics",
      },
      "image": {
        "url": "https://waifu.mock/nekos",
      },
      "timestamp": "2022-01-15T17:15:20.066Z",
      "title": "Classic Neko Style",
    },
  ],
}
`;
